src/microengine_utils/scanner.py

```diff
@@ -0,0 +1,165 @@
+    stderr: 'asyncio.StreamReader' = asyncio.subprocess.PIPE,
+    stdin: 'asyncio.StreamReader' = asyncio.subprocess.DEVNULL,
+    check: 'bool' = False,
+    timeout: 'int' = 15,
+) -> 'Tuple[ExitCodeT, Text, ...]':
+    """Run an engine filescan `cmd`, timing out after `timeout` seconds"""
+    try:
+        proc = await asyncio.subprocess.create_subprocess_exec(
+            *cmd,
+            stdout=stdout,
+            stderr=stderr,
+            stdin=stdin,
+        )
+        streams = await asyncio.wait_for(proc.communicate(), timeout=timeout)
+        if check and proc.returncode != 0:
+            raise CalledProcessScanError
+        return (
+            proc.returncode,
+            *(s.decode(errors='ignore') if isinstance(s, bytes) else str(s or '') for s in streams)
+        )
+    except (BrokenPipeError, ConnectionResetError):  # noqa
+        proc.kill()
+        raise CalledProcessScanError
+    except asyncio.TimeoutError as e:
+        proc.kill()
+        raise TimeoutScanError
+
+
+def collect_scan(verbose: 'bool' = False) -> 'Callable[..., ScanResult]':
+    """Decorator for `async_scan` to automatically handle errors and boilerplate scanner metadata
+
+    - Record and send timing data to Datadog
+    - Read the `ScanResult`'s fields to automatically figure out which metrics should be collected
+    - Merges `ScanResult` `metadata` with boilerplate scanner information from `EngineInfo`
+    """
+    statsd = datadog.statsd
+
+    def wrapper(scan_fn: 'Callable'):
+        @functools.wraps(scan_fn)
+        async def driver(
+            self,
+            guid: 'str',
+            artifact_type: 'ArtifactType',
+            content: 'bytes',
+            metadata: 'Dict',
+            chain: 'Optional[str]',
+        ):
+            start = time.time()
+            tags = []
+
+            # e.g 'type:file' or 'type:url'
+            tags.append('type:%s' % ArtifactType.to_string(artifact_type))
+
+            try:
+                scan = await scan_fn(guid, artifact_type, content, metadata, chain)
+            except BaseScanError as e:
+                # e.g 'scan_error:fileskippedscanerror'
+                tags.append('scan_error:{e.event_name}')
+                statsd.increment(SCAN_FAIL, tags=tags)
+                # If we encountered a scan error, we still return a scan result with `scan_error`
+                # documenting the error which occurred
+                scan = ScanResult(
+                    bit=False,
+                    verdict=False,
+                    metadata=Verdict().set_malware_name('').set_extra('scan_error', e.event_name),
+                )
+            else:
+                # No result
+                if scan.bit is False:
+                    if verbose:
+                        statsd.increment(SCAN_NO_RESULT, tags=tags)
+                # Verdict reported
+                elif scan.bit is True and (scan.verdict is True or scan.verdict is False):
```

Comment:
the right side of the `and` has a test for both True and False, what other options are there? Is `None` an option for that value?

Reply:
`verdict` defaults to `False`

Reply:
Though I think we were discussing a move to allow bit True, verdict None to create a suspicious label.